contrib/lib/Wcl/DemosMotif/xmailbox.man
.\";;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
.\"
.\" File:         xmailbox.man
.\" RCS:          $Header$
.\" Description:  man page template
.\" Author:	  Andrew Peebles
.\" Created:      1-5-91
.\" Modified:
.\" Language:     Text
.\" Package:      
.\"
.\";;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
.TH XMAILBOX 1 "" "Internal Tools"
.ad b
.SH NAME
xmailbox \- a general purpose mailbox monitor
.SH SYNOPSIS
.br
\fBxmailbox\fP [ \-options ] [ \-toolkit options ]
.SH DESCRIPTION
\fIXmailbox\fP is a Widget Creation Library (Wcl) based Motif application
that is used primarily to monitor your mailbox file, telling you when new
mail arrives and who its from.  It is not a mail reader.  The user interface
is entirely programmable by the user via Wcl resource files.  The name
of the mailbox file and the commands used to gather From and Subject lines
are resources and can be changed from thier default values to monitor
other types of files (new users added to /etc/passwd, new machines added
to /etc/hosts, etc).

See the xmailbox 'manual' XMailbox.README_FIRST for a better description
of xmailbox features and examples of how to run it.  See the Wcl manuals
to learn how to write Wcl resource files.

The options below can be specified to xmailbox either via resources in
a resource file or by command line options.  The resource name is shown
in the first column, with the equivelent command line option name shown
in the next column surrounded by ().

.SH OPTIONS
.br
.LP
.B mailbox (-mailbox)
.sp
	The full pathname and file name of the mailbox file 
	to monitor.  Typically this will be something like 
	/usr/mail/peebles but may be any file that you wish 
	to monitor changes to.
.br
.LP
.B wcResources (-resources)
.sp
	The name of the resource file containing the Wcl 
	resources that define the user interface.
.br
.LP
.B hideNoMail (-hide)
.sp
	Boolean value indicating whether to hide the xmailbox
	window (pop it down) when there is no new mail, or
 	when the mailbox file becomes empty.
.br
.br
.LP
.B includeSubject (-subject)
.sp
	Boolean value indicating whether to apply the
	subjectCmd (see below) to the mailbox file when
	new mail arrives.  Its sometimes useful to set this
	to False when you write a custom fromCmd filter to
	look at the mailbox file.
.br
.br
.LP
.B flash (-flash)
.sp
	Boolean value.  When True causes the xmailbox window 
	to flash when new mail arrives in the mailbox file.
.br
.br
.LP
.B flashTimeout (-ftimeout)
.sp
	Microseconds between flashs when flash=True.
.br
.br
.LP
.B flashNumber (-fnumber)
.sp
	Number of times to flash if flash=True.
.br
.br
.LP
.B beep (-beep)
.sp
	Boolean value.  When True, causes xmailbox to beep
	when new mail has arrived.
.br
.br
.LP
.B timeout (-timeout)
.sp
	Number of seconds between scans of the mailbox 
	file to check if new mail has arrived.  This value
	defaults to 60 seconds.
.br
.br
.LP
.B nomailboxWarningMessage (-nowarning)
.sp
	Boolean value.  If True, you get no warning message 
	if the mailbox file does not exsist.  If False you
	get a warning message and xmailbox terminates.
	Some unix systems remove the mail spool files when
	no mail exsists for the user.
.br
.br
.LP
.B fromCmd (-fromCmd)
.sp
	The name of the command to use to scan the mailbox
	file for "From" lines.  This command will be passed
	a single argument, the mailbox file specified with
	the mailbox resource, and is expected to produce
	output to stdout.  It defaults to 'grep "From:"'.
.br
.br
.LP
.B subjectCmd (-subjectCmd)
.sp
	Same as the fromCmd resource, but used to scan for
	"Subject" lines.  It defaults to 'grep "Subject:"'.
	This command is not performed if the resoure
	includeSubject is False.
.br

.SH More on from and subject Commands
The fromCmd and subjectCmd resources are used to format the data that
is shown in the xmailbox window when the mailbox file is updated with
more data.  When xmailbox scans the mailbox file (controlled by the
timeout resource) it first checks to see if the mailbox file is zero length.
If it is, xmailbox pops itself down (becomes invisible) if hideNoMail is
True.  If the mailbox file is not zero lenght and its length is different
than the last time is was scanned, then the fromCmd is invoked with
the mailbox file name passed as a command line argument.  The standard
output of the fromCmd is read back into xmailbox.  Only lines that are
new (not already displayed in the xmailbox window) are appended to the
bottom of the display window.  If includeSubject is true, the subjectCmd
is also invoked, and its output displayed next to the output of the
fromCmd.

Let us build a simple /etc/hosts monitor.  It will look at /etc/hosts and
display any new machines added to the hosts file while xmailbox is running.
First we'll specify the 'mailbox file' to look at:

XMailbox.mailbox:	/etc/hosts

And a simple fromCmd that returns the first machine name after the host
address in the hosts file:

XMailbox.fromCmd:	awk -f '{print $2}'

We don't want the subjectCmd, because all we're after is new machine names:

XMailbox.includeSubject:False

When you first invoke xmailbox with these resources, it will list all the
machines currently in the hosts file.  You can then use the ClearText
command from the interface to rub out the contents of the xmailbox window.
Now whenever new machines are added to the hosts file, only those new
machines will be displayed in the xmailbox window.

Another example, this time using the subjectCmd, is to keep an eye on the
passwd file.  You might wish to be notified (using xmailbox) when a new
user has been added to the passwd file by sysadm, and to see both the
new user's name and home directory.  Perhaps you'd like to send new users
a welcome message when they first join your company.  You'd use something
like this:

XMailbox.mailbox:	/etc/passwd
.br
XMailbox.fromCmd:	awk -F':' '{print $1}'
.br
XMailbox.subjectCmd:	awk -F':' '{print $6}'
.br
XMailbox.includeSubject:True

I'm sure you can come up with even more interesting uses for xmailbox.
Each of these uses can each have thier own unique user interface via
the wcResources resource, so they all look like seperate applications
and are customized to thier own special tasks.

.SH AUTHOR
Andrew Peebles, Mips Computer Systems
.SH SEE\ ALSO
Wcl manual, XMailbox.README_FIRST



contrib/lib/Wcl/DemosMotif/xmailbox.c
/***************************************************************************
 *
 * FILE NAME		: xmailbox.c
 *
 * AUTHOR		: Andrew Peebles
 *
 * DESCRIPTION		: Simple utility to monitor mail box
 *
 * VERSIONS		: %W%
 *
 ***************************************************************************/

/*
 * standard includes
 */
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <Xm/Xm.h>

#include <X11/Shell.h>
#include <Wc/WcCreate.h>

static void ClearTextACT();
static void HideMailACT();
static void ReScanACT();

static XtActionsRec	MYactions[] = {
  {"ClearTextACT",	ClearTextACT},
  {"HideMailACT",	HideMailACT},
  {"ReScanACT",		ReScanACT},
};

/*
 * resources
 */
typedef struct _defs {
  char		*mailbox;
  Boolean	hide_nomail;
  Boolean	include_subject;
  Boolean	flash;
  int		flash_timeout;
  int		flash_number;
  Boolean	beep;
  char		*mri_file;
  int		timeout;
  char		*from_cmd;
  char		*subject_cmd;
  Boolean	nomailboxWarningMessage;
} Defs, *DefsPtr;
static Defs defaults;

static	int 	timeout;
static	Boolean	hide_nomail, include_subject, flash, beep;
static	char	*mailbox, *mri_file;
static	int	flash_timeout, flash_number;
static	char	*from_cmd, *subject_cmd;

static XtResource resources[] = {
  {"mailbox", "Mailbox",
     XmRString, sizeof (char *),
     XtOffset (DefsPtr, mailbox),
     XmRString,
     NULL},

  {"hideNoMail", "HideNoMail",
     XmRBoolean, sizeof (Boolean),
     XtOffset (DefsPtr, hide_nomail),
     XmRString,
     "False"},

  {"nomailboxWarningMessage", "NomailboxWarningMessage",
     XmRBoolean, sizeof (Boolean),
     XtOffset (DefsPtr, nomailboxWarningMessage),
     XmRString,
     "True"},

  {"flash", "Flash",
     XmRBoolean, sizeof (Boolean),
     XtOffset (DefsPtr, flash),
     XmRString,
     "False"},

  {"beep", "Beep",
     XmRBoolean, sizeof (Boolean),
     XtOffset (DefsPtr, beep),
     XmRString,
     "False"},

  {"includeSubject", "IncludeSubject",
     XmRBoolean, sizeof (Boolean),
     XtOffset (DefsPtr, include_subject),
     XmRString,
     "False"},

  {"timeout", "Timeout",
     XmRInt, sizeof (int),
     XtOffset (DefsPtr, timeout),
     XmRString,
     "60"},

  {"flashTimeout", "FlashTimeout",
     XmRInt, sizeof (int),
     XtOffset (DefsPtr, flash_timeout),
     XmRString,
     "100"},

  {"flashNumber", "FlashNumber",
     XmRInt, sizeof (int),
     XtOffset (DefsPtr, flash_number),
     XmRString,
     "10"},

  {"wcResources", "WcResources",
     XmRString, sizeof (char *),
     XtOffset (DefsPtr, mri_file),
     XmRString,
     "/user/cadroot/lib/Mri/XMailbox.res"},

  {"fromCmd", "FromCmd",
     XmRString, sizeof (char *),
     XtOffset (DefsPtr, from_cmd),
     XmRString,
     "grep \"From:\""},

  {"subjectCmd", "SubjectCmd",
     XmRString, sizeof (char *),
     XtOffset (DefsPtr, subject_cmd),
     XmRString,
     "grep \"Subject:\""},

};

/*
 * Command line equivelents
 */
static XrmOptionDescRec cmd_options[] =
{
  {"-mailbox",		".mailbox",	XrmoptionSepArg,	NULL},
  {"-resources",	".wcResources",	XrmoptionSepArg,	"/user/cadroot/lib/Mri/XMailbox.res"},
  {"-hide",		".hideNoMail",	XrmoptionNoArg,		"True"},
  {"-subject",		".includeSubject", XrmoptionNoArg,	"True"},
  {"-flash",		".flash",	XrmoptionNoArg,		"True"},
  {"-beep",		".beep",	XrmoptionNoArg,		"True"},
  {"-timeout",		".timeout",	XrmoptionSepArg,	"60"},
  {"-ftimeout",		".flashTimeout", XrmoptionSepArg,	"100"},
  {"-fnumber",		".flashNumber",	XrmoptionSepArg,	"10"},
  {"-nowarning",	".nomailboxWarningMessage", XrmoptionSepArg, "False"},
  {"-fromCmd",		".fromCmd",	XrmoptionSepArg,	NULL},
  {"-subjectCmd",	".subjectCmd",	XrmoptionSepArg,	NULL},
};


/*
 * global vars
 */
static	XtAppContext 	app;
static	Widget		text, shell;
static	Boolean		UP = False;
static  Boolean		DoAutoRaise = False;
static  Widget		topshell;
static	int		last_size;
static	int		flash_count = 0;
static	int		box_size = 0;

#define FLASH_MAX	6

char	*mystrdup();
static  void ClearText();
static	void FlashEvent();
static	Boolean ScanMBox();

main (argc, argv)
int	argc;
char	*argv;
{
  Widget	appshell;
  void		TimeOutEvent();
  char		*resfile;
  Display	*display;
  int		n;
  Arg		args[5];
  int		mfp;

  topshell = XtInitialize ("xmailbox", "XMailbox",
			   cmd_options, XtNumber(cmd_options),
			   &argc, argv);
#if 0
  appshell = XtAppCreateShell ("xmailbox", "XMailbox",
			       applicationShellWidgetClass,
                               XtDisplay(topshell), NULL, 0);
#endif
  n=0;
  XtSetArg (args[n], XmNtransient, True); n++;
  shell = XtCreatePopupShell ("xmailbox",
			      topLevelShellWidgetClass,	
                              topshell, args, n);

  app = XtWidgetToApplicationContext (topshell);

  XtAppAddActions (app, MYactions, XtNumber(MYactions));

  XtGetApplicationResources (topshell,
                             &defaults,
                             resources, XtNumber(resources),
                             NULL, 0);

  timeout = defaults.timeout;
  hide_nomail = defaults.hide_nomail;
  include_subject = defaults.include_subject;
  flash = defaults.flash;
  beep = defaults.beep;
  flash_timeout = defaults.flash_timeout;
  flash_number = defaults.flash_number;
  if (flash_number % 2) ++flash_number;
  if (defaults.mailbox == NULL) {
    fprintf (stderr,"MUST specify a mailbox resource!\n");
    exit (1);
  }
  mailbox = mystrdup (defaults.mailbox);

  if ((mfp = open (mailbox, O_RDONLY)) == -1) {
    if (defaults.nomailboxWarningMessage) {
      fprintf (stderr, "Can't open mailbox: %s\n",mailbox);
    }
  }

#ifdef DEBUG
  printf ("timeout = %d seconds\n", timeout);
  printf ("hide_nomail = %s\n",((hide_nomail) ? "True" : "False"));
  printf ("include_subject = %s\n", ((include_subject) ? "True" : "False"));
  printf ("mailbox = %s\n",mailbox);
#endif

  RegisterApplicationCallbacks (app);
  MriRegisterMotif(app);

  resfile = (char *) malloc (strlen(defaults.mri_file) + 1);
  strcpy (resfile, defaults.mri_file);
  WcLoadResourceFileCB (shell, resfile, NULL);

  if( WcWidgetCreation(shell) != 0 )
    return(1);

  XtRealizeWidget (shell);

#ifdef DEBUG
  ListWidgetNames (appshell);
#endif

  XtAppAddTimeOut (app,
                   1000 * 1,	/* get to it fast for the first time */
                   TimeOutEvent,
                   NULL);

  XtPopup (shell, XtGrabNone);
  UP = True;

  XtMainLoop();
}

static Boolean first_up = True;

void
TimeOutEvent (type, id)
int             type;
XtIntervalId    *id;
{
  int FLASH = False;

#ifdef DEBUG
  printf ("timeout ...\n");
#endif

  FLASH = ScanMBox();

  XtAppAddTimeOut (app,
                   1000 * timeout,
                   TimeOutEvent,
                   NULL);

  if (FLASH) {
    FLASH = False;
    flash_count = 0;
    XtAppAddTimeOut (app,
		     flash_timeout,
		     FlashEvent,
		     NULL);
  }

  if (first_up) first_up = False;

}

static Boolean
ScanMBox()
{
  struct stat stats;
  int	FLASH = False;

  /*
   * Try to stat mailbox, see if any mail
   */
  stats.st_size = 0;
  if ((stat (mailbox, &stats) == -1) &&
      (defaults.nomailboxWarningMessage)) {
    Error ("Can't stat (open) mailbox");
    return (True);
  }
  else if ((stats.st_size != 0) && (stats.st_size > last_size)) {

    last_size = stats.st_size;

    if (DealWithText(True)) {
      if (UP) {
	if (DoAutoRaise) {
	  XRaiseWindow (XtDisplay (shell), XtWindow (shell));
	}
      } else {
	XtPopup (shell, XtGrabNone);
	UP = True;
      }
    
      if (beep) {
	XBell (XtDisplay(shell), 99);
      }

      if (flash) {
	FLASH = True;
      }
    }

  }
  else if (stats.st_size == 0) {
    if (hide_nomail && UP && !first_up) {
      XtPopdown (shell);
      UP = False;
    } else {
      if (DoAutoRaise)
	XLowerWindow (XtDisplay (shell), XtWindow (shell));
    }
    box_size = 0;
    last_size = 0;
    ClearText (NULL, NULL, NULL);
  }
  else if (stats.st_size < last_size) {
    last_size = stats.st_size;
    box_size = 0;
    DealWithText (False);	/* recalcualtes box_size */
  }

  return (FLASH);
}

static void
FlashEvent (type, id)
int             type;
XtIntervalId    *id;
{
  Arg	args[5];
  int	n;
  Pixel	fg, bg;
  int	pos;

  if (flash_count == flash_number)
    return;

  pos = XmTextGetLastPosition (text);

  n=0;
  XtSetArg (args[n], XmNbackground, &bg); n++;
  XtSetArg (args[n], XmNforeground, &fg); n++;
  XtGetValues (text, args, n);

  n=0;
  XtSetArg (args[n], XmNbackground, fg); n++;
  XtSetArg (args[n], XmNforeground, bg); n++;
  XtSetValues (text, args, n);

  XmTextSetInsertionPosition (text, pos);

  ++flash_count;
  XtAppAddTimeOut (app,
		   flash_timeout,
		   FlashEvent,
		   NULL);
}

/*
 * User accessabible Callback routines
 */

static void
RegisterText (widget, widgetName, unused)
Widget  widget;
char*   widgetName;
caddr_t unused;
{
  Widget        cp;
  char          clean[128];

  WcCleanName (widgetName, clean);
  text = WcFullNameToWidget (widget, clean);
}

static void
ClearText (widget, widgetName, unused)
Widget  widget;
char*   widgetName;
caddr_t unused;
{
  XmTextSetString (text, NULL);
}

static void
HideMail (widget, widgetName, unused)
Widget  widget;
char*   widgetName;
caddr_t unused;
{
  XtPopdown (shell);
  UP = False;
}

static void
AutoRaise (widget, widgetName, unused)
{
  DoAutoRaise = !DoAutoRaise;
}

static void
ReScan (widget, widgetName, unused)
Widget  widget;
char*   widgetName;
caddr_t unused;
{
  box_size = 0;
  last_size = 0;
  ClearText (NULL, NULL, NULL);
  ScanMBox();
}

RegisterApplicationCallbacks (app)
XtAppContext app;
{
#define RCALL( name, func ) WcRegisterCallback ( app, name, func, NULL );

  RCALL ( "RegisterText",  	RegisterText       );
  RCALL ( "ClearTextCB",	ClearText	   );
  RCALL ( "HideMailCB",		HideMail	   );
  RCALL ( "ReScanCB",		ReScan		   );
  /*
   * and for backward compatability
   */
  RCALL ( "ClearText",		ClearText	   );
  RCALL ( "HideMail",		HideMail	   );
  RCALL ( "ReScan",		ReScan		   );
  RCALL ( "AutoRaise",		AutoRaise	   );
}

DealWithText(displayit)
Boolean displayit;
{
  int pos, THIS_POS;
  char	cmd[128];
  FILE	*popen(), *fp, *sp;
  char	from[180], subject[180];
  int	index;
  char	*tok;
  int	ret = False;

  sprintf (cmd,"%s %s",defaults.from_cmd,mailbox);

  if ((fp = popen (cmd, "r")) == NULL) {
    Error ("Problems reading From from mailbox");
    return (0);
  }

  if (include_subject) {

    sprintf (cmd,"%s %s",defaults.subject_cmd,mailbox);

    if ((sp = popen (cmd, "r")) == NULL) {
      Error ("Problems reading Subject from mailbox");
      pclose (fp);
      return (0);
    }

  }

  index = 0;
  while (fgets (from, 180, fp) != NULL) {
    if (include_subject) {
      fgets (subject, 180, sp);
    }

    if (index++ < box_size) {
      continue;
    }

#ifdef DEBUG
    printf (">> %s ...\n", from);
#endif

    ++box_size;

    if (!displayit)
      continue;

    ret = True;

    if (!include_subject) {
      if (from[strlen(from)-1] != '\n') {
	from[strlen(from)-1] = '\n';
      }
      tok = from;
      THIS_POS = 
      pos = XmTextGetLastPosition (text);
      XmTextReplace (text, pos, pos, tok);
    }
    else {
      if (from[strlen(from)-1] == '\n') {
	from[strlen(from)-1] = '\0';
      }
      if (subject[strlen(subject)-1] != '\n') {
	subject[strlen(subject)-1] = '\n';
      }
      tok = from;
      THIS_POS = 
      pos = XmTextGetLastPosition (text);
      XmTextReplace (text, pos, pos, tok);

      tok = subject;
      pos = XmTextGetLastPosition (text);
      XmTextReplace (text, pos, pos, tok);
    }
    bzero (from, 80);
    bzero (subject, 80);
  }

  XmTextSetInsertionPosition (text, THIS_POS);

  pclose(fp);
  if (include_subject)
    pclose(sp);

  return (ret);
}

Error (msg) 
char *msg;
{
  fprintf (stderr, "xmailbox Warning: %s\n", msg);
}

char *
mystrdup (str)
char	*str;
{
  char	*ret = (char *) malloc (strlen(str) + 1);
  strcpy (ret, str);
  return (ret);
}

/*
 * action functions - they mearly call thier callback equivelents
 */
static void
ClearTextACT()
{
  ClearText();
}

static void
HideMailACT()
{
  HideMail();
}

static void
ReScanACT()
{
  ReScan();
}

